### PR TITLE
ID-261 CSRF Mitigation

### PR DESCRIPTION
**Pull request by tlangs**
Remediates https://broadworkbench.atlassian.net/browse/ID-261

Bond needs to implement CSRF protection. To do this, we will provide a random `nonce` in the `state` parameter of the url returned from the `authorization-url` endpoint in Bond. When a client uses that url and gets back an OAuth code, the client must provide that same `nonce` in the state in order for Bond to successfully respond to the `oauthcode` request. 

This is a breaking change, meaning Terra-UI will need to be updated. This cannot merge until Terra-UI forwards state along to bond as a part of the `oauthcode` call.

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary. 
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: bond_app/oauth2_state_store.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import base64
+import json
+import secrets
+from collections import namedtuple
+from google.cloud import ndb
+
+# Information associated with a tokens for refreshing service account credentials.
+OAuth2StateInfo = namedtuple('OAuth2StateInfo', ['nonce'])
```
Comment (Ghost-in-a-Jar) [inline bond_app/oauth2_state_store.py]: There's finally a better way to do this type of thing:
https://realpython.com/python-data-classes/
ex:
```
from dataclasses import dataclass
@dataclass
class OAuth2StateInfo:
    nonce: str
    
```

File: bond_app/oauth2_state_store.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    def save(self, user_id, provider, nonce):
+        """
+        Persists a RefreshToken by creating a new entity or updating an existing entity with the same id
+        :param user_id
+        :param provider:
+        :param nonce: random value for csrf protection
+        """
+        ouath2_nonce = OAuth2State(key=OAuth2StateStore._oauth2_state_store_key(user_id, provider), nonce=nonce)
+        ouath2_nonce.put()
```
Comment (Ghost-in-a-Jar) [inline bond_app/oauth2_state_store.py]: what does this do?

Comment (Ghost-in-a-Jar) [reply bond_app/oauth2_state_store.py]: oh also typo heh

Comment (Ghost-in-a-Jar) [reply bond_app/oauth2_state_store.py]: oh i see this persists it in the datastore, what is the underlying db we are using here?

Comment (tlangs) [reply bond_app/oauth2_state_store.py]: There is no underlying _real_ DB. We're using https://cloud.google.com/datastore/

Comment (Ghost-in-a-Jar) [reply bond_app/oauth2_state_store.py]: ok gotcha 👍

File: bond_app/oauth2_state_store.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    def validate_and_delete(self, user_id, provider_name, nonce) -> bool:
+        key = OAuth2StateStore._oauth2_state_store_key(user_id, provider_name)
+        oauth2_state = key.get()
+        is_valid = False
+        if oauth2_state:
+            if oauth2_state.nonce == nonce:
+                is_valid = True
+            key.delete()
+        return is_valid
```
Comment (Ghost-in-a-Jar) [inline bond_app/oauth2_state_store.py]: nitpicky: Id say something like this is slightly more readable:
```
oauth2_state = key.get()
if oauth2_state is None:
    return False
else:
    key.delete()
    return oauth2_state.nonce == nonce
```

File: bond_app/oauth2_state_store.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    def state_with_nonce(self, state):
+        if not state:
+            decoded_state = {}
+        else:
+            decoded_state = json.loads(base64.b64decode(state))
+        nonce = secrets.token_urlsafe()
+        decoded_state_with_nonce = {**decoded_state, 'nonce': nonce}
+        return base64.b64encode(json.dumps(decoded_state_with_nonce).encode('utf-8')), nonce
```
Comment (Ghost-in-a-Jar) [inline bond_app/oauth2_state_store.py]: Just for the sake of making stack traces easier to read id say split the json dumps and encoding on different lines

File: tests/unit/fake_oauth2_state_store.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    def state_with_nonce(self, state):
+        if not state:
+            decoded_state = {}
+        else:
+            decoded_state = json.loads(base64.b64decode(state))
+        nonce = self.test_nonce
+        decoded_state_with_nonce = {**decoded_state, 'nonce': nonce}
+        return base64.b64encode(json.dumps(decoded_state_with_nonce).encode('utf-8')), nonce
+
```
Comment (Ghost-in-a-Jar) [inline tests/unit/fake_oauth2_state_store.py:53]: same 2 comments as in the actual class

**Review by Ghost-in-a-Jar — APPROVED** [review]
I think this is good and esp if you have tested it manually locally -- one nice to have though would be a last test (and more of these in the repo in general) that used a flask test client to actually stand up the app and call endpoints. https://flask.palletsprojects.com/en/1.1.x/testing/#the-first-test

**Comment by tlangs** [comment]
I think @gpolumbo-broad might have words to say about adding more "Integration" or "end-to-end" tests to the repo. 😅

File: bond_app/oauth2_state_store.py (hunk 1 above)
Comment (tlangs) [reply bond_app/oauth2_state_store.py]: oh cool! Like Records! Or Case Classes!

**Comment by tlangs** [comment]
@Ghost-in-a-Jar Turns out we actually have "automation" tests that test a deployed branch of Bond upon PR.

File: bond_app/oauth2_state_store.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class OAuth2StateStore:
+    """
+    Stores OAuth2 State nonces for csrf protection.
+    """
+
+    def save(self, user_id, provider, nonce):
+        """
+        Persists a RefreshToken by creating a new entity or updating an existing entity with the same id
```
Comment (gpolumbo-broad) [inline bond_app/oauth2_state_store.py]: RefreshToken?

File: swagger/api-docs.yaml [hunk 7]
```diff
@@ -74,7 +74,6 @@ paths:
       summary: Builds an OAuth authorization URL that a user must use to initiate the OAuth dance.
```
Comment (andy7i) [inline swagger/api-docs.yaml]:
```suggestion
      summary: Builds an OAuth authorization URL that a user must use to initiate the OAuth dance. Creates a 1 time use link. There can only be 1 active link per user per provider. The last link created will invalidate any prior links.
```

**Review by andy7i — COMMENTED** [review]
partial review since Greg is almost done reviewing

**Comment by tlangs** [comment]
jenkins retest

**Review by gpolumbo-broad — APPROVED** [review]

**Pull request merged**